README
This is C::DynaLib, a Perl 5 module for calling C functions.
Example:

  use C::DynaLib;
  $libm = new C::DynaLib("-lm");
  $sinh = $libm->DeclareSub("sinh", "d", "d");
  print "The hyperbolic sine of 3 is ", &$sinh(3), "\n";


	***********************************************************
	***   THIS CODE CONTAINS SYSTEM DEPENDENCIES.           ***
	***   IT WILL NOT WORK ON ALL COMPILERS, MACHINES, OR   ***
	***   OPERATING SYSTEMS.        USE AT YOUR OWN RISK.   ***
	***********************************************************


	***********************************************************
	***   A BETTER WAY TO INTERFACE TO C CODE IS WITH XS.   ***
	***   PLEASE READ THE "perlxs" DOCUMENTATION BEFORE     ***
	***   MAKING EXTENSIVE USE OF THIS MODULE.              ***
	***********************************************************


The module has been tested successfully on iX86, Sparc, and Alpha
machines.  However, there are probably compilers for even these chips
which will require some porting. Please report any problems at rt.cpan.org

Included are the following packages:

  C::DynaLib            - call a C function from Perl
  C::DynaLib::Callback  - convert a perl sub into a usable
                                     C function pointer
  C::DynaLib::Struct    - yet another tool for handling C
                                     compound data structures
  Asm                   - unsupported whimsical i386 CPU interface
                                     (not installed by default)

To build this module, run the usual

  perl Makefile.PL
  make test

Various things can go wrong.  First, your perl should be of the dynamic
variety.  Be sure you answered "y" when Configure asked, "Do you wish
to use dynamic loading?".  Otherwise, you will still (I think) be able
to create callback pointers and convert a function pointer into a Perl
sub, but you won't be able to do dynamic loading.

The biggest issue in porting this module is your system's function
call interface.  If Makefile.PL prints that it is using `hack30', it
is a sign that your system is not yet supported.  Another sign of
trouble is when `make test' prints `not ok' or `Illegal memory
operation'.  Further documentation in pod format is in DynaLib.pm.

The callback feature has sometimes caused grief in building the
module.  If Makefile.PL writes a makefile but you can't get it to
build, try `perl Makefile.PL CALLBACKS=0' followed by `make test',
and report of your success or failure.

COPYRIGHT

The files named in MANIFEST, accompanying this file, are Copyright (c)
1997, 2000 by John Tobey.  All rights reserved.  They may be modified,
distributed and used under the same conditions as Perl itself.  Of
course, there is no express or implied warranty.  See the file README
in the top-level Perl source directory for more information.

Enjoy!
-John
jtobey@john-edwin-tobey.org
and Reini Urban <rurban@cpan.org>


Changes
Revision history for Perl extension C::DynaLib 
(formerly ExtUtils::DynaLib, very formerly CLib).

0.61  2010-12-21 rurban
	Added experimental calling conventions cdecl3, cdecl6, 
          cdecltr ("typed return") and dummy stdcall (NYI)
	Fix t/win32.t
	Better detect gcc
	Improve perl Makefile.PL Verbose
	added experimental script/hparse.pl header parser
        added sample script/cpuid,pl
        removed Test::Simple dependency
	added repository http://github.com/rurban/c-dynalib
	fix Quad issues introduced with 0.59
	added dl_unload_file as used by dlopen/win32 platforms 
          (most fortunately)
	fix msvc6 activeperl 5.10 against malloc/free/realloc defines
        fix C::DynaLib::Struct::Parse

0.60  2010-04-20 23:55:12 rurban
        Improve library search and windows resolution:
          know about their libc, try dllimport on importlibs.

0.59  2010-04-18 15:55:36 rurban
	Added C::DynaLib::Struct->Parse with the help of 
	  Convert::Binary::C struct parser.
        Make Convert::Binary::C optional
        Generate C::DynaLib::PerlTypes
        Added testreverse check.
        Added is_big_endian check.
        Added check_ieee_fp check.
        Add -g on -DDEBUGGING

0.58  2008-06-29 12:59:11 rurban
	removed C::DynaLib::Struct::VERSION
	work on failing win32 test

0.57  2008-01-18 12:34:43 rurban
	Fix optimize for non-gcc, non-msvc
        define $C::DynaLib::decl to "cdecl", "hack30", ...
        skip hack30 pow() double test
	reversed Changes

0.56  2005-09-20 18:34:43 - 2007-12-26 22:38:11 rurban
	Maintainer changed to Reini Urban <rurban@cpan.org>
	Allow to build under cygwin and gcc >= 3.4
	Added script/pl2exe.pl, moved tests to t/

0.55  Fri Feb 11 19:54:34 EST 2000
	Allow to build with recent perls.

0.53  Sat Oct 25 12:47:32  1997
	Improved error checking and reporting.  Function calls occur
in the sub declarer's package.  Second arg to C::DynaLib->new will be
passed along to DynaLoader::dl_load_file.

0.52  Sat Oct 18 18:41:42  1997
	Some efforts toward portability, especially among Sparc
compilers.

0.51  Fri Oct  3 23:55:27  1997
	Renamed as C::DynaLib.  Fixed Makefile.PL bug that resulted
in NUL chars in DynaLib.c.  Errors in loading libs, finding symbols,
and allocating callbacks are no longer fatal.

0.50  Sat Sep 27 01:35:55  1997
	Much nicification and many added arg/return types.  Rewrote
a lot of the guts.

0.31  Thu Aug 28 23:42  1997
	Digital Unix on the Alpha is supported, at least for gcc,
thanks to help from Achim Bohnet <ach@mpe.mpg.de>.  Added PTR_TYPE.

0.30  Mon Aug 11 00:58:55  1997
	Added testcall configuration utility to find system
dependencies for cdecl convention.  Rewrote some docs.

0.22  Wed Jul 30 01:12 1997
	Renamed methods to avoid possible future keyword conflict:
declare_sub became DeclareSub, poke became Poke, etc.  Added
ExtUtils::DynaLib::Struct package.  Got it to work on Windows 95 with
Borland C++ (but probably not repeatably ;-) )

0.21  Sat Jul 12 02:01:12  1997
	die works with callbacks (except in certain complex cases).
Added a few bits & pieces.
Sun's cc is now supported (thanks to lvirden@cas.org).

0.20  Mon Jul  7 23:42:00  1997
	Support for callbacks.  Major and minor bugfixes.

0.13-0.14  Nicified for CPAN.

0.12  Sun Jun 22 10:34:21  1997
	Minor stuff.  Renamed to ExtUtils::DynaLib.

0.11  Thu Jun 19 23:43:29  1997
	Added sparc convention.

0.10  Wed Jun 18 13:31:07 EDT 1997
	Got `hack30' to return ints correctly on sparc/solaris.
Made compatible with perl 5.003.

0.04  Tue Jun 17 04:15:57  1997
	Generalized the default calling convention; constructor
accepts filename; declare_sub accepts function pointer; added ``hack30''
calling convention

0.03  Sun Jun 15 22:56:26  1997
	Fixed problem with Perl prototypes.

0.02  Sat Jun 14 19:22:36  1997
	First release, only "cdecl" convention supported.

0.01  Thu Jun 12 01:17:41 1997
	- original version; created by h2xs 1.18


Makefile.PL
use ExtUtils::MakeMaker;

use strict;
require 5.002;
use Config;

my(@convention, $convention,
   $num_callbacks, %cflags, $object, $devnull,
   $conv_xs, $cbfunc_c, $cdecl_h, $postamble,
   $is_gcc, $is_msvc, $is_borland, $is_dynamic,
   $is_win32, $Verbose,
   $stack_reserve, $stack_adjust, $stack_align,
   $do_reverse);

my $Cpp; # run testcall through -E -P (might break, just for debugging)
$conv_xs = "conv.xsi";
$cbfunc_c = "cbfunc.c";
$cdecl_h = "cdecl.h";
$object = '$(BASEEXT)$(OBJ_EXT)';

$is_gcc = $Config{cc} =~ /gcc/i && $Config{gccversion} >= 2;
$is_msvc = ($^O eq 'MSWin32' and $Config{cc} =~ /^cl/i);
$is_borland = $Config{cc} =~ /\bbcc/i;
$is_dynamic = ($Config{usedl} eq 'define');
$is_win32   = $^O =~ /MSWin32|cygwin/;
use subs qw(write_conv write_cbfunc write_cdecl_h is_big_endian check_ieee_fp);

$postamble = "
clean::
	\$(RM_F) $conv_xs $cbfunc_c *.core *.stackdump

$conv_xs: $0 \$(CONFIGDEP)

$cbfunc_c: $0 \$(CONFIGDEP)

DynaLib.c: $conv_xs
";


$is_dynamic or warn <<STATIC;

*** NOTE
*** According to $INC{"Config.pm"},
*** this perl does not know how to use dynamic loading.  The test
*** program for this module will fail, and you will not be able to
*** invoke functions in dynamic libraries.  If you need this feature,
*** you have to rebuild perl.  Choose "y" when Configure asks, "Do
*** you wish to use dynamic loading?".

STATIC


@convention = ();
%cflags = ();
for (@ARGV) {
    /^DECL=(.*)$/ and push @convention, split(",", $1);
    /^CALLBACKS=(\d+)$/ and $num_callbacks = $1;
    /^(-D.*)(?:\=(.*))?$/ and $cflags{$1} = $2;
    /^Verbose$/i and $Verbose = 1;
    /^Cpp$/i and $Cpp = 1;
    /^STACK_RESERVE=(\d+)$/ and $stack_reserve = $1;
    /^STACK_ADJUST=(\d+)$/ and $stack_adjust = $1;
    /^STACK_ALIGN=(\d+)$/ and $stack_align = $1;
}

# Appease MakeMaker:
@ARGV = grep { !/^(DECL=|CALLBACKS=\d+$|-D.|STACK_\w+=\d+)/ } @ARGV;

my $ccflags = $Config{ccflags};
my $archname = $Config{archname};
# cc can also be gcc
if (!$is_gcc and $Config{cc} =~ /^cc/) {
    my $test = `$Config{cc} -dumpversion`;
    $is_gcc = 1 if $test and $test eq $Config{gccversion}."\n";
    $is_gcc = 1 if $test and !$is_gcc and $test =~ /^[34]./;
}
# TODO: support more fastcall derivates (ia64/x86_64: first six in regs, rest in args)
unless (@convention) {
    for (
	 #[ '^i?[x3-6]86', sub {'cdecl'} ],
	 [ '^sun4-', sub {'sparc'} ],
	 [ 'sparc', sub {'sparc'} ],
	 [ '(alpha|axp)', 
	   sub {
	       unless ($is_gcc) {
		   $postamble .= "\nalpha-cc\$(OBJ_EXT): alpha-cc.s\n"
		       . "\t\$(CC) -c alpha-cc.s -o \$\@\n"
		       . "\n#alpha-cc.s: alpha-cc.c\n"
		       . "#\tgcc -O2 -S alpha-cc.c -o \$\@\n";
		   $object .= " alpha-cc.o";
	       }
	       'alpha'} ],
	 [ 'win32', sub {('cdecl','stdcall')} ],
	 [ '^cygwin', sub {my $CDECL_ADJUST = -32; ('cdecl','stdcall')} ],
	 [ '', sub { () } ],
	 )
    { @convention = &{$_->[1]}, last if $archname =~ /$_->[0]/i }
}
# when? see 75b18f7c2d75e
# freebsd7 gcc3.4.6 cdecl ok   one_by_one=0,do_adjust=0,stack_reserve=0,do_reverse=0,arg_align=4
# freebsd7 gcc4.2.1 cdecl ok   one_by_one=0,do_adjust=0,stack_reserve=0,do_reverse=0,arg_align=4
#---
# freebsd7 gcc4.2.1 cdecl3 ok  one_by_one=0,do_adjust=-16,stack_reserve=3,do_reverse=0,arg_align=4
# cygwin7  gcc4.3.4 cdecl3 ok  one_by_one=0,do_adjust=-16,stack_reserve=3,do_reverse=0,arg_align=4
# mingw    gcc3.4.5 cdecl fail one_by_one=0,do_adjust=-16,stack_reserve=3,do_reverse=0,arg_align=4
# msvc6    cl 12    cdecl ok   one_by_one=1,do_adjust=0,stack_reserve=0,do_reverse=0,arg_align=4
# debian64 gcc4.4.4 cdecl6 ok  one_by_one=0,do_adjust=-16,stack_reserve=6,do_reverse=0,arg_align=8

# Disable cc optimization
my $optimize = $Config{optimize};
if ($is_msvc) {
    $optimize = '/Od';
    $ccflags =~ s|[-/]O\d ||;
} elsif ($Config{optimize} =~ /[\-\/]O\w/)  {
    ; #$optimize =~ s/O\w/O0/;  # changing to -O0 turned out to be contra-productive
}
if ($is_gcc and $ccflags =~ /-DDEBUGGING/) {
    $optimize .= " -g";
}
#if ($is_gcc and $ccflags !~ /omit-frame-pointer/) {
#    # but it does not work as expected on linux+win32. use the same as perl
#    $optimize .= " -fomit-frame-pointer";
#}

WriteMakefile(
    'NAME'	=> 'C::DynaLib',
    'VERSION_FROM' => 'lib/C/DynaLib.pm',
    'DEFINE'	=> '',
    'OBJECT'	=> $object,

    # Don't let MakeMaker set up a dependency loop.
    # cdecl.h depends on testcall.o, not the other way around!
    'H'         => [],
    'C'         => ['DynaLib.c'],
    'INC'	=> '-I.',
    'OPTIMIZE'	=> $optimize,

    'PREREQ_PM' => {
        'Convert::Binary::C'   => 0.70, # 0.70 structs only
	($is_gcc ? ('GCC::TranslationUnit' => "1.00") : ()),
        'Test'         => 0,
        'Test::More'   => 0,
    },
    ($] >= 5.005 ?
     ('AUTHOR' => 'John Tobey <John.Tobey@gmail.com>, Reini Urban <rurban@cpan.org>',
      'ABSTRACT_FROM' => 'lib/C/DynaLib.pm') : ()),
    ($ExtUtils::MakeMaker::VERSION gt '6.46' ?
    ('META_MERGE'
     => {recommends =>
	 {
	  "GCC::TranslationUnit" => "1.01",
	  "Convert::Binary::C"   => "0.74",
	 },
	 resources =>
	 {
	  license     => 'http://dev.perl.org/licenses/',
	  repository  => 'http://github.com/rurban/c-dynalib',
	 },
	}
    ) : ()),
    clean => { FILES =>
 	       "*.core *.stackdump *.exe test*.ilk test*.obj test*.pdb test*.pdb vc*.pdb ".
               "*.err *.bak *.o dll.base dll.exp"
             },
);


sub is_big_endian
{
    my $byteorder = $Config{byteorder}
      || unpack( "a*", pack "L", 0x34333231 );
    die "Native byte order ($byteorder) not supported!\n"
      if   $byteorder ne '1234'     and $byteorder ne '4321'
       and $byteorder ne '12345678' and $byteorder ne '87654321';
    $byteorder eq '4321' or $byteorder eq '87654321';
}

sub check_ieee_fp
{
    my @test = (
    {
     value  => '-1.0',
     double => pack( 'C*', 0xBF, 0xF0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 ),
     single => pack( 'C*', 0xBF, 0x80, 0x00, 0x00 ),
    },
    {
     value  => '0.0',
     double => pack( 'C*', 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 ),
     single => pack( 'C*', 0x00, 0x00, 0x00, 0x00 ),
    },
    {
     value  => '0.4',
     double => pack( 'C*', 0x3F, 0xD9, 0x99, 0x99, 0x99, 0x99, 0x99, 0x9A ),
     single => pack( 'C*', 0x3E, 0xCC, 0xCC, 0xCD ),
    },
    {
     value  => '1.0',
     double => pack( 'C*', 0x3F, 0xF0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 ),
     single => pack( 'C*', 0x3F, 0x80, 0x00, 0x00 ),
    },
    {
     value  => '3.1415926535',
     double => pack( 'C*', 0x40, 0x09, 0x21, 0xFB, 0x54, 0x41, 0x17, 0x44 ),
     single => pack( 'C*', 0x40, 0x49, 0x0F, 0xDB ),
    },
    {
     value  => '1.220703125e-4',
     double => pack( 'C*', 0x3F, 0x20, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 ),
     single => pack( 'C*', 0x39, 0x00, 0x00, 0x00 ),
    },
  );
  my @fail;
  for my $t ( @test ) {
      my $s = pack 'f', $t->{value};
      my $d = pack 'd', $t->{value};
      unless( &is_big_endian ) { for( $s, $d ) { $_ = reverse $_ } }

      $s eq $t->{single} or push @fail,
        { value => $t->{value}, check => 'single', expected => $t->{single}, got => $s };

      $d eq $t->{double} or push @fail,
        { value => $t->{value}, check => 'double', expected => $t->{double}, got => $d };
  }
  return @fail;
}

sub pretest {
    my $self = shift;
    my @candidate;
    my $cleanup = sub {
	unlink ("testtest.c", "testtest.txt", "testtest$self->{EXE_EXT}",
		"testtest$self->{OBJ_EXT}");
    };
    my $cant = sub {
	&$cleanup;
	print "I don't know how run the test program,\n",
	      "so I'll try to find a default configuration.\n"
	    if $Verbose;
	return undef;
    };
    my $try = sub {
	open HELLO, "testtest.txt" or return undef;
	my $hello = <HELLO>;
	close HELLO;
	if ($hello =~ /hello, world/) {
	    &$cleanup;
	    return 1;
	}
	return 0;
    };

    &$cleanup;
    unless (open PROG, ">testtest.c") {
	print ("Can't write testtest.c\n");
	return &$cant;
    }
    print PROG <<'PROG';
#define main notmain
#include <EXTERN.h>
#include <perl.h>
#undef main
#undef fprintf
#undef fopen
#undef fclose
#include <stdio.h>
int main()
{
    FILE *fp = fopen("testtest.txt", "w");
    if (fp == NULL) {
	return 1;
    }
    fprintf(fp, "hello, world!\n");
    fclose(fp);
    return 0;
}
PROG
    close PROG;

    @candidate = ();
    $devnull = $^O eq 'MSWin32' ? "> NUL" : ">/dev/null 2>&1";
    my $cmd1 = "$Config{cc} $ccflags $optimize -DNARF -I$self->{PERL_INC} testtest.c";
    push @candidate, "$cmd1 -o testtest$self->{EXE_EXT} $devnull"
	unless $Verbose;
    push @candidate, "$cmd1 -otesttest$self->{EXE_EXT} $devnull"
	unless $Verbose;
    push @candidate, "$cmd1 -o testtest$self->{EXE_EXT}";
    push @candidate, "$cmd1 -otesttest$self->{EXE_EXT}";

    while ($self->{how_to_compile} = shift (@candidate)) {
	unlink "testtest$self->{EXE_EXT}";
	print "$self->{how_to_compile}\n" if $Verbose;
	system ($self->{how_to_compile});
	last if $? == 0 && -x "testtest$self->{EXE_EXT}";
    }
    return &$cant unless $self->{how_to_compile};

    @candidate = ();
    push @candidate, "./testtest$self->{EXE_EXT} $devnull"
	if !$is_msvc and !$Verbose;
    push @candidate, "./testtest$self->{EXE_EXT}"
      if !$is_msvc;
    push @candidate, "testtest$self->{EXE_EXT} $devnull"
	unless $Verbose;
    push @candidate, "testtest$self->{EXE_EXT}";
    push @candidate, "run testtest$self->{EXE_EXT}";

    unlink ("testtest.txt");
    while ($self->{how_to_run} = shift (@candidate)) {
	print "$self->{how_to_run}\n" if $Verbose;
	system ($self->{how_to_run});
	$? == 0 && &$try and return 1;
    }
    return &$cant;
}

sub guess_cdecl_h {
    my $self = shift;
    open CONFIG, ">$cdecl_h" or die "can't write $cdecl_h";

    $do_reverse = 0;
    print "Testing stack layout...\n" if $Verbose;
    $self->{how_to_compile} =~ s/testtest/testreverse/g;
    $self->{how_to_run} =~ s/testtest/testreverse/g;
    my $defines;
    for $defines ("-DINCLUDE_ALLOCA", "", "-DINCLUDE_MALLOC") {
	my $cmd = $self->{how_to_compile};
	$defines .= " -g" if $is_gcc;
	$defines .= " -DVERBOSE" if $Verbose;
	$cmd =~ s/-DNARF/$defines/g;
	unlink ("testreverse$self->{EXE_EXT}", $cdecl_h);
	print "$cmd\n" if $Verbose;
	system ($cmd);
	if ($? == 0 && -x "testreverse$self->{EXE_EXT}") {
	    $cmd = $self->{how_to_run};
	    print "$cmd\n" if $Verbose;
	    system ($cmd);
	    if ($? == 0) {
	      $do_reverse = 0;
	    } else {
	      $do_reverse = 1;
	    }
	    print "CDECL_REVERSE = $do_reverse\n" if $Verbose;
	}
    }
    my $define_if_not = sub {
	my ($macro, $def) = @_;
	return "#ifndef $macro\n#define $macro $def\n#endif\n\n";
    };
    print CONFIG <<CONFIG;
/*
 * $cdecl_h generated by $0.  Do not edit this file, edit $0.
 */
CONFIG
    print CONFIG "#include <alloca.h>\n"
	if $self->{CC} =~ /\bcc$/;
    print CONFIG "#include <malloc.h>\n"
	if $is_borland;
    print CONFIG (&$define_if_not("CDECL_ONE_BY_ONE",
		  (($archname =~ /win32/i && ! $is_borland)
		   || $is_gcc) ? 1 : 0));
    print CONFIG (&$define_if_not("CDECL_ADJUST",
		  ($is_borland ? -12 : 0)));
    print CONFIG (&$define_if_not("CDECL_REVERSE", $do_reverse));

    close CONFIG;
}

sub make_postamble {
    my $self = shift;

    $postamble .= "\nDynaLib\$(OBJ_EXT): DynaLib.c $cbfunc_c"
      . " @{[ map { \"$_.c\" } @convention ]}\n";

    ! @convention || grep { $_ eq "cdecl" } @convention
	or return $postamble;

    print "Writing $cdecl_h\n";
    if (write_cdecl_h($self)) {
	unless (@convention) {
	    # check cdecl/cdecl3/cdecl6
	    local $/;
	    open F, "<", $cdecl_h;
	    my $file = <F>;
	    my ($cdecl3) = $file =~ /CDECL_STACK_RESERVE (\d)/;
	    my ($reverse) = $file =~ /CDECL_REVERSE (\d)/;
	    close F;
	    if ($cdecl3 == 3) {
		@convention = ('cdecl3');
	    } elsif ($cdecl3 == 6) {
		@convention = ('cdecl6');
	    } else {
		@convention = ('cdecl');
	    }
	    if ($reverse or $is_win32) {
		push @convention, 'stdcall';
		print "Using conventions: @convention\n";
	    }
	}
    } elsif (@convention) {
	print "Can't figure out this system.  I'll have to guess.\n"
	    if $Verbose;
	guess_cdecl_h($self);
	if ($do_reverse) {
	    push @convention, 'stdcall';
	    print "Using conventions: @convention\n";
	}
	elsif ($is_win32) {
	    push @convention, 'stdcall', 'hack30';
	    print "Using conventions: @convention\n";
	}
    } else {
	print <<WARN;
***
*** WARNING
***
*** I can not figure out the correct way to pass arguments to a C function
*** on this system.  This may be due to porting issues, a perl installation
*** problem, or any number of things.  Maybe 'perl Makefile.PL Verbose'
*** will shed some light.
***
*** I will use the "hack30" calling convention, which may work some or most
*** of the time for integers alike. Or it may crash your programs. A better 
*** solution would be to add support for your systems calling convention.
***
*** See perldoc lib/C/DynaLib.pm for a discussion of "hack30".
***
WARN
	@convention = ('hack30');
	return $postamble;
    }

    $postamble .= "
clean::
	\$(RM_F) testreverse\$(EXE_EXT) testreverse\$(OBJ_EXT) testcall\$(EXE_EXT) testcall\$(OBJ_EXT) $cdecl_h

DynaLib\$(OBJ_EXT): $cdecl_h ";
    $postamble .= join(" ", map { $_.".c"} @convention);
    $postamble .= "

$cdecl_h: $0 \$(CONFIGDEP) testcall\$(EXE_EXT)
	$self->{how_to_run}

testcall\$(EXE_EXT) : testcall.c
	\$(CC) `\$(PERL) -MExtUtils::Embed -e ccopts` \$(OPTIMIZE) testcall.c -o\$\@

testreverse\$(EXE_EXT) : testreverse.c
	\$(CC) `\$(PERL) -MExtUtils::Embed -e ccopts` \$(OPTIMIZE) testreverse.c -o\$\@
";
}

sub MY::postamble {
    my $self = shift;
    my $postamble = make_postamble($self);

    print "Using calling convention(s): @convention\n"
      if $Verbose;
    for (@convention) {
      # $cflags{"-DDYNALIB_DECL=\\\"$_\\\""} = undef;
      $cflags{"-DDYNALIB_USE_$_"} = undef;
    }

    print "Default calling convention: $convention[0]\n"
      if $Verbose;
    $cflags{"-DDYNALIB_DEFAULT_CONV"} = "\\\"$convention[0]\\\"";

    $num_callbacks = 4 unless defined($num_callbacks);
    print "Maximum number of callbacks: $num_callbacks\n"
      if $Verbose;
    $cflags{"-DDYNALIB_NUM_CALLBACKS"} = $num_callbacks;

    my $defines = "\nDEFINE =";
    for (sort keys %cflags) {
	$defines .= " $_";
	$defines .= "=$cflags{$_}" if defined $cflags{$_};
    }
    $postamble .= $defines;
    print "Additional definitions: $defines\n"
      if $Verbose and scalar(keys %cflags)>1;
    write_conv();
    write_cbfunc();
    parse_perl_types();

    return $postamble;
}

sub write_cdecl_h {
    my $self = shift;

    print "Testing how to compile and run a program...\n"
	if $Verbose;
    pretest($self) or return undef;

    print "Testing how to pass args to a function...\n"
	if $Verbose;
    $self->{how_to_compile} =~ s/testtest/testcall/g;
    $self->{how_to_run} =~ s/testtest/testcall/g;
    my $defines;
    for $defines ("-DINCLUDE_ALLOCA", "", "-DINCLUDE_MALLOC") {
	my $cmd = $self->{how_to_compile};
	$cmd =~ s/-DNARF/$defines/g;
	$cmd .= " -g" if $is_gcc;
	$cmd .= " -DVERBOSE" if $Verbose;
	if ($is_gcc and $Cpp) { # expand DO_CALL for gdb
	    $cmd =~ s/-o\s+testcall/-o testcall.E.c -E -P/;
	    print "$cmd\n" if $Verbose;
	    system ($cmd);
	    if ($? == 0 && -f "testcall.E.c") {
		# add \n to DO_CALL and fix linenumbers for gdb
		open IN, "<", "testcall.E.c";
		open OUT, ">", "testcall.E1.c";
		while (<IN>){
		    # line numbers stripped with -P, but keep this code.
		    if (/^# (\d+) "testcall.E1.c"$/) {
			if ($1 > 50) {
			    s/^.*$//;
			    next;
			}
		    }
		    s/^# (\d+) "testcall.c"/# $. "testcall.E1.c"/;
		    if (/ char \*arg; int i; void \*d1/) {
			s/(;|\{|\}) /$1\n\t/g;
		    }
		    print OUT;
		}
		close IN;
		close OUT;
		$cmd =~ s/testcall.c -o testcall.E.c -E -P/testcall.E1.c -o testcall/;
	    }
	}
      AGAIN:
	unlink ("testcall$self->{EXE_EXT}", $cdecl_h);
	print "$cmd\n" if $Verbose;
	system ($cmd);
	if ($? == 0 && -x "testcall$self->{EXE_EXT}") {
	    my $runcmd = $self->{how_to_run};
	    print "$runcmd\n" if $Verbose;
	    system ($runcmd);
	    if ($? == 0 && -e $cdecl_h) {
		print "Succeeded.\n" if $Verbose;
		return 1;
	    }
	    # compiled okay and executable: no need to try the two
	    # other options no header and malloc
	    if ($cmd =~ / -DRESERVE/) {
		last;
	    } else {
		print "Try again with -DRESERVE\n" if $Verbose;
		$cmd .= " -DRESERVE";
		goto AGAIN;
	    }
	}
    }
    return undef;
}

sub write_conv {

    # Write conv.xsi, to be included in DynaLib.xs

    open XS, ">$conv_xs"
      or die "Can't write file \"$conv_xs\": $!\n";
    print "Writing $conv_xs\n";

    print XS <<XS;
#
# $conv_xs generated by $0.  Don't edit this file, edit $0.
#
XS

#
# conv.xsi XS definition for the "glue" function that calls C.
#
for $convention (@convention) {
    print XS <<XS;

void
${convention}_call_packed(symref, ret_type, ...)
	void *		symref
	char *		ret_type
	PROTOTYPE: \$\$\@
	PPCODE:
	{
	  SV *sv;
#ifdef HAS_QUAD
	  Quad_t aquad;
	  unsigned Quad_t auquad;
#endif

	  if (*ret_type != '\\0') {
	    sv = sv_newmortal();
	  }
	  switch (*ret_type) {
	  case '\\0' :
	    (void) ${convention}_CALL(symref, int);
	    XSRETURN_EMPTY;
	  case 'i' :
	    sv_setiv(sv, (IV) ${convention}_CALL(symref, int));
	    break;
	  case 'l' :
	    sv_setiv(sv, (IV) ${convention}_CALL(symref, I32));
	    break;
	  case 's' :
	    sv_setiv(sv, (IV) ${convention}_CALL(symref, I16));
	    break;
	  case 'c' :
	    sv_setiv(sv, (IV) ${convention}_CALL(symref, char));
	    break;
	  case 'I' :
	    sv_setuv(sv, (UV) ${convention}_CALL(symref, uint));
	    break;
	  case 'L' :
	    sv_setuv(sv, (UV) ${convention}_CALL(symref, U32));
	    break;
	  case 'S' :
	    sv_setuv(sv, (UV) ${convention}_CALL(symref, U16));
	    break;
	  case 'C' :
	    sv_setuv(sv, (UV) ${convention}_CALL(symref, uchar));
	    break;
#ifdef HAS_QUAD
	  case 'q' :
	    aquad = ${convention}_CALL(symref, Quad_t);
	    if (aquad >= IV_MIN && aquad <= IV_MAX)
	      sv_setiv(sv, (IV)aquad);
	    else
	      sv_setnv(sv, (double)aquad);
	    break;
	  case 'Q' :
	    aquad = ${convention}_CALL(symref, Uquad_t);
	    if (aquad <= UV_MAX)
	      sv_setuv(sv, (UV)auquad);
	    else
	      sv_setnv(sv, (double)auquad);
	    break;
#endif
	  case 'f' :
            sv_setnv(sv, (double) ${convention}_CALL(symref, float));
	    break;
	  case 'd' :
	    sv_setnv(sv, ${convention}_CALL(symref, double));
	    break;
	  case 'Z' :
	  case 'p' :
	    sv_setpv(sv, (char *) ${convention}_CALL(symref, char*));
	    break;
	  case 'P' :
	    sv_setpvn(sv, (unsigned char *) ${convention}_CALL(symref, char*),
			  atoi(&ret_type[1]));
	    break;
	  default :
	    croak("Unsupported function return type: '%c'", *ret_type);
	  }
	  XPUSHs(sv);
	  XSRETURN(1);
	}
XS
}
close XS;
}

sub write_cbfunc {
    my ($i);

    # Write cbfunc.c, to be included in DynaLib.xs

    open FUNCS, ">$cbfunc_c"
      or die "Can't write file \"$cbfunc_c\": $!\n";
    print "Writing $cbfunc_c\n";

    print FUNCS <<FUNCS;
/*
 * $cbfunc_c generated by $0.  Don't edit this file, edit $0.
 */
FUNCS
    #
    # The callback functions.
    #
    for $i (0 .. $num_callbacks - 1) {
	print FUNCS <<FUNCS;

static long
#ifdef I_STDARG
_cb_func$i(void * first, ...)
#else
_cb_func$i(first, va_alist)
void * first;
va_dcl
#endif
{
  va_list ap;
  long result;

#ifdef I_STDARG
  va_start(ap,first);
#else
  va_start(ap);
#endif
  result = cb_call_sub($i, first, ap);
  va_end(ap);
  return result;
}
FUNCS
    }

    #
    # Array of callback entry pointers.
    #
    print FUNCS "\nstatic const cb_callback cb_arr[DYNALIB_NUM_CALLBACKS] = {\n";
    for $i (0 .. $num_callbacks - 1) {
	print FUNCS "\t_cb_func$i,\n";
    }
    print FUNCS "};\n";
}

sub parse_perl_types
{
    unless (eval "require Convert::Binary::C;") {
	print "Warning: Convert::Binary::C not installed. PerlTypes not generated.\n";
	return;
    }
    Convert::Binary::C->import;
    use Data::Dumper;
    my $byteorder = is_big_endian();
    my $c = new Convert::Binary::C('ByteOrder' => $byteorder ? 'BigEndian' : 'LittleEndian');
    $c->Include(["$Config{archlib}/CORE", '/usr/include', $Config{incpath},
		 $Config{locincpth}, @{$c->Include}]);
    eval { $c->parse("#include <EXTERN.h>\n#include <perl.h>"); };
    my $cfg = $c->configure;
    my $dump = Data::Dumper->Dump([$cfg], ['PerlTypes']);
    my $fname = "lib/C/DynaLib/PerlTypes.pm";
    open XS, "> $fname"
      or die "Can't write file \"$fname\": $!\n";
    print "Adding Convert::Binary::C configuration to $fname\n";
    print XS <<XS;
# PerlTypes.pm generated by Makefile.PL.  Don't edit this file, edit Makefile.PL.
# From Convert::Binary::C \$c->parse("#include <EXTERN.h>\n#include <perl.h>")->configure

package C::DynaLib::PerlTypes;

$dump;

1;
__END__
XS
    # eval C::DynaLib::PERLTYPES => %C::DynaLib::Struct::PerlTypes;
    close XS;
}

# Local Variables:
#   cperl-indent-level: 4
# End
